lib/rules/disallow-curly-braces.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Disallows curly braces after statements.
+ *
+ * Type: `Array` or `Boolean`
+ *
+ * Values: Array of quoted keywords or `true` to disallow curly braces after the following keywords:
+ *
+ * #### Example
+ *
+ * ```js
+ * "disallowCurlyBraces": [
+ *     "if",
+ *     "else",
+ *     "while",
+ *     "for",
+ *     "do",
+ *     "with"
+ * ]
+ * ```
+ *
+ * ##### Valid
+ *
+ * ```js
+ * if (x) x++;
+ * ```
+ *
+ * ##### Invalid
+ *
+ * ```js
+ * if (x) {
+ *     x++;
+ * }
+ * ```
+ */
+
+var assert = require('assert');
+var defaultKeywords = require('../utils').curlyBracedKeywords;
+
+module.exports = function() {};
+
+module.exports.prototype = {
+
+    configure: function(statementTypes) {
+        assert(
+            Array.isArray(statementTypes) || statementTypes === true,
+            this.getOptionName() + ' option requires array or true value'
+        );
+
+        if (statementTypes === true) {
+            statementTypes = defaultKeywords;
+        }
+
+        this._typeIndex = {};
+        statementTypes.forEach(function(type) {
+            this._typeIndex[type] = true;
+        }.bind(this));
+    },
+
+    getOptionName: function() {
+        return 'disallowCurlyBraces';
+    },
+
+    check: function(file, errors) {
+
+        function isBlockStatement(node) {
+            return node && node.type === 'BlockStatement';
+        }
+
+        function isMultiline(node) {
+            return node.loc.end.line - node.loc.start.line > 2;
+        }
+
+        function addError(typeString, entity) {
+            errors.add(
+                typeString + ' statement with extra curly braces',
+                entity.loc.start.line,
+                entity.loc.start.column
+            );
+        }
+
+        function checkBody(type, typeString) {
+            file.iterateNodesByType(type, function(node) {
+                if (isMultiline(node)) {
+                    return;
+                }
+                if (isBlockStatement(node.body)) {
+                    addError(typeString, node);
+                }
+            });
+        }
+
+        var typeIndex = this._typeIndex;
+
+        if (typeIndex['if'] || typeIndex['else']) {
+            file.iterateNodesByType('IfStatement', function(node) {
+                if (typeIndex['if'] && isBlockStatement(node.consequent)) {
+                    if (isMultiline(node.consequent)) {
```

Comment:
Move this check into `isBlockStatement` rather than calling it after every time.